applications/IgaApplication/custom_conditions/base_discrete_condition.cpp

```diff
@@ -0,0 +1,344 @@
+        }
+
+        KRATOS_CATCH("")
+    }
+
+    void BaseDiscreteCondition::GetDofList(
+        DofsVectorType& rElementalDofList,
+        ProcessInfo& rCurrentProcessInfo
+    )
+    {
+        KRATOS_TRY;
+
+        KRATOS_ERROR_IF(DofsPerNode() != 3) << "Calling BaseDiscreteCondition::GetDofList. Derived Condition does not has 3 DoFs." << std::endl;
+
+        const int number_of_control_points = NumberOfNodes();
+
+        rElementalDofList.resize(0);
+        rElementalDofList.reserve(3 * number_of_control_points);
+
+        for (unsigned int i = 0; i < number_of_control_points; ++i) {
+            rElementalDofList.push_back(GetGeometry()[i].pGetDof(DISPLACEMENT_X));
+            rElementalDofList.push_back(GetGeometry()[i].pGetDof(DISPLACEMENT_Y));
+            rElementalDofList.push_back(GetGeometry()[i].pGetDof(DISPLACEMENT_Z));
+        }
+
+        KRATOS_CATCH("")
+    };
+
+    void BaseDiscreteCondition::CalculateRightHandSide(
+        VectorType& rRightHandSideVector,
+        ProcessInfo& rCurrentProcessInfo)
+    {
+        const std::size_t number_of_dofs = NumberOfDofs();
+
+        MatrixType left_hand_side_matrix = Matrix(0, 0);
+
+        if (rRightHandSideVector.size() != number_of_dofs) {
+            rRightHandSideVector.resize(number_of_dofs);
+        }
+
+        CalculateAll(left_hand_side_matrix, rRightHandSideVector,
+            rCurrentProcessInfo, false, true);
+    }
+
+    void BaseDiscreteCondition::CalculateLeftHandSide(
+        MatrixType& rLeftHandSideMatrix,
+        ProcessInfo& rCurrentProcessInfo)
+    {
+        const std::size_t number_of_dofs = NumberOfDofs();
+
+        if (rLeftHandSideMatrix.size1() != number_of_dofs
+            || rLeftHandSideMatrix.size2() != number_of_dofs) {
+            rLeftHandSideMatrix.resize(number_of_dofs, number_of_dofs);
+        }
+
+        VectorType right_hand_side_vector = Vector(0);
+
+        CalculateAll(rLeftHandSideMatrix, right_hand_side_vector,
+            rCurrentProcessInfo, true, false);
+    }
+
+    void BaseDiscreteCondition::CalculateLocalSystem(
+        MatrixType& rLeftHandSideMatrix,
+        VectorType& rRightHandSideVector,
+        ProcessInfo& rCurrentProcessInfo)
+    {
+        const std::size_t number_of_dofs = NumberOfDofs();
+
+        if (rLeftHandSideMatrix.size1() != number_of_dofs
+            || rLeftHandSideMatrix.size2() != number_of_dofs) {
+            rLeftHandSideMatrix.resize(number_of_dofs, number_of_dofs);
+        }
+
+        if (rRightHandSideVector.size() != number_of_dofs) {
+            rRightHandSideVector.resize(number_of_dofs);
+        }
+
+        CalculateAll(rLeftHandSideMatrix, rRightHandSideVector,
+            rCurrentProcessInfo, true, true);
+    }
+
+    void BaseDiscreteCondition::Calculate(
+        const Variable<array_1d<double, 3>>& rVariable,
+        array_1d<double, 3>& rOutput,
+        const ProcessInfo& rCurrentProcessInfo)
+    {
+        if (rVariable == VELOCITY) {
```

Comment:
I am pretty sure that we can simplify this methods a lot. If I have some time I will try to implement it to the `IgaBaseElement`